### ISSUE TITLE
CUDA not available when calling ` backwards` before using CUDA

### ISSUE THREAD

**Issue by shaltielshmid** [dotnet/TorchSharp#1198]
When I use TorchSharp and I perform a gradient calculation before loading in the CUDA binaries, then TorchSharp no longer recognizes that CUDA is available.

For example:
```cs
var lin = torch.nn.Linear(10, 1, false);
lin.forward(torch.rand(10)).backward();
Console.WriteLine(torch.cuda.is_available()); // False
```

On the other hand, if I call a CUDA function before the `backward()` call, it will work:

```cs
Console.WriteLine(torch.cuda.is_available()); // True
var lin = torch.nn.Linear(10, 1, false);
lin.forward(torch.rand(10)).backward();
Console.WriteLine(torch.cuda.is_available()); // True
```

I've traced this issue back to LibTorch, over [here](https://github.com/pytorch/pytorch/blob/HEAD/aten/src/ATen/detail/CUDAHooksInterface.cpp#L28-L29). The exact quote is:
```cs
// NB: The once_flag here implies that if you try to call any CUDA
// functionality before libATen_cuda.so is loaded, CUDA is permanently
// disabled for that copy of ATen.
```

@NiklasGustafsson How do you think this should be handled?

In PyTorch I'm pretty sure that they load in the CUDA binaries with the first `import torch` call. This would be easy to add to TorchSharp, but is that the desired behavior?

**Comment by NiklasGustafsson** [comment]
Hmm... I think we just need to place a call to 'cuda.is_available()' on some path that is guaranteed to be run before any real work is done. A static initializer block (is that what they're called? I use them so rarely) somewhere in the `torch` static class, perhaps? What do you think, @shaltielshmid?

**Comment by shaltielshmid** [comment]
I was actually thinking we can make torch.CUDA be a field (like torch.CPU, torch.META, etc.). This would require adding an if and replacing it with null, if it doesn't exist. 
Alternatively, a static initializer is also a great solution. It would guarantee that cuda would be initialized before any use of torch operations.

**Comment by NiklasGustafsson** [comment]
We need something that isn't going to throw an exception if the CUDA backend isn't there. Calling cuda.is_available() is safe that way.

**Comment by shaltielshmid** [comment]
I was thinking of this:
```cs
public static Device CUDA = TryInitializeDeviceType(DeviceType.CUDA) ? new Device(DeviceType.CUDA, -1) : null;
```

But I'm not set on it. Should I create a PR with the static initializer?

**Comment by NiklasGustafsson** [comment]
This looks cleaner:

```C#
public static Device CUDA = cuda.is_available() ? new Device(DeviceType.CUDA, -1) : null;
```

**Comment by NiklasGustafsson** [comment]
> But I'm not set on it. Should I create a PR with the static initializer?

As long as the solution does not throw in the absence of a CUDA backend, I'm okay with it. The field should be read-only, right?

**Comment by shaltielshmid** [comment]
Yes! We should probably also set torch.CPU and torch.META to be readonly as well.

**Comment by NiklasGustafsson** [comment]
I'm not sure if it initializes the CUDA backend. Don't know that off the top of my head.

**Comment by shaltielshmid** [comment]
It does - that's where I got the TryInitialize idea :)

**Comment by NiklasGustafsson** [comment]
Great. Then calling is_available is more readable, IMO

**Comment by shaltielshmid** [comment]
Great! I'll have something ready soon
